samples/nrf9160/azure_iot_hub/src/main.c

```diff
@@ -0,0 +1,460 @@
+		 */
+		k_delayed_work_submit_to_queue(&application_work_q,
+					       &send_event_work, K_NO_WAIT);
+		break;
+	case AZURE_IOT_HUB_EVT_DATA_RECEIVED:
+		printk("AZURE_IOT_HUB_EVT_DATA_RECEIVED\n");
+		printf("Received payload: %.*s\n",
+		       evt->data.msg.len, evt->data.msg.ptr);
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_CONNECTING:
+		printk("AZURE_IOT_HUB_EVT_DPS_CONNECTING\n");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_REGISTERING:
+		printk("AZURE_IOT_HUB_EVT_DPS_REGISTERING\n");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_DONE:
+		printk("AZURE_IOT_HUB_EVT_DPS_DONE\n");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_FAILED:
+		printk("AZURE_IOT_HUB_EVT_DPS_FAILED\n");
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RECEIVED:
+		printk("AZURE_IOT_HUB_EVT_TWIN_RECEIVED\n");
+		event_interval_apply(event_interval_get(evt->data.msg.ptr));
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED:
+		printk("AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED\n");
+		on_evt_twin_desired(evt->data.msg.ptr, evt->data.msg.len);
+		break;
+	case AZURE_IOT_HUB_EVT_DIRECT_METHOD:
+		printk("AZURE_IOT_HUB_EVT_DIRECT_METHOD\n");
+		on_evt_direct_method(&evt->data.method);
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS:
+		printk("AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS, ID: %d\n",
+		       evt->data.result.rid);
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RESULT_FAIL:
+		printk("AZURE_IOT_HUB_EVT_TWIN_RESULT_FAIL, ID %d, status %d\n",
+			evt->data.result.rid, evt->data.result.status);
+		break;
+	case AZURE_IOT_HUB_EVT_PUBACK:
+		printk("AZURE_IOT_HUB_EVT_PUBACK\n");
+		break;
+	default:
+		printk("Unknown Azure IoT Hub event type: %d\n", evt->type);
+		break;
+	}
+}
+
+static void send_event(struct k_work *work)
+{
+	int err;
+	static char buf[60];
+	ssize_t len;
+	struct azure_iot_hub_data msg = {
+		.topic.type = AZURE_IOT_HUB_TOPIC_EVENT,
+		.ptr = buf,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+	};
+
+	len = snprintf(buf, sizeof(buf),
+		       "{\"temperature\":%d.%d,\"timestamp\":%d}",
+		       25, k_uptime_get_32() % 10, k_uptime_get_32());
+	if ((len < 0) || (len > sizeof(buf))) {
+		printk("Failed to populate event buffer\n");
+		goto exit;
+	}
+
+	msg.len = len;
+
+	printk("Sending event:\n%s\n", buf);
+
+	err = azure_iot_hub_send(&msg);
+	if (err) {
+		printk("Failed to send event\n");
+		goto exit;
+	}
+
+	printk("Event was successfully sent\n");
+exit:
+	if (atomic_get(&event_interval) <= 0) {
+		printk("The event reporting stops, interval is set to %d\n",
+		       atomic_get(&event_interval));
+		return;
+	}
+
+	printk("Next event will be sent in %d seconds\n", event_interval);
+	k_delayed_work_submit_to_queue(&application_work_q, &send_event_work,
+				       K_SECONDS(event_interval));
+}
+
+static void direct_method_handler(struct k_work *work)
+{
+	int err;
+	static const char ret[] = "{\"it\":\"worked\"}";
+	struct azure_iot_hub_result result = {
+		.rid = method_data.request_id,
+		.status = 200,
+		.payload = ret,
+		.payload_len = sizeof(ret) - 1,
```

Comment:
nit: maybe use strlen() and remove the -1. Since the payload is a null-terminated string. Includes less magic.

Reply:
I can change it, but usually before the compile time `sizeof` for constants